File: apps/chat/src/kick-webhook.ts
import { createVerify } from "node:crypto";

import { erro, parseJson, validate } from "@lebedevna/neverthrow-utils";
import { ok, Result, safeTry, type Result as NeverthrowResult } from "neverthrow";
import { z } from "zod";

import type { ChatEventBroker } from "./chat-application.js";
import type { ChatStore } from "./store.js";

const WEBHOOK_CLOCK_SKEW_MS = 5 * 60 * 1_000;

const KickIdentitySchema = z.object({
  user_id: z.union([z.string(), z.number()]).transform(String),
  username: z.string().min(1),
});

const KickMessageSchema = z.object({
  message_id: z.string().min(1),
  broadcaster: KickIdentitySchema,
  sender: KickIdentitySchema,
  content: z.string(),
  created_at: z.string().datetime(),
});

type KickWebhookError = Readonly<{
  type: "invalid kick webhook" | "unknown kick source";
  detail: string;
  cause?: unknown;
}>;

function requiredHeader(headers: Headers, name: string) {
  const value = headers.get(name);
  return value !== null
    ? ok(value)
    : erro<KickWebhookError>({ type: "invalid kick webhook", detail: `${name} is missing` });
}

function verifySignature(
  publicKey: string,
  messageId: string,
  timestamp: string,
  body: string,
  signature: string,
): NeverthrowResult<void, KickWebhookError> {
  const occurredAt = Date.parse(timestamp);
  if (!Number.isFinite(occurredAt) || Math.abs(Date.now() - occurredAt) > WEBHOOK_CLOCK_SKEW_MS) {
    return erro({ type: "invalid kick webhook", detail: "Kick webhook timestamp is stale" });
  }
  return Result.fromThrowable(
    () => {
      const verifier = createVerify("RSA-SHA256");
      verifier.update(`${messageId}.${timestamp}.${body}`);
      verifier.end();
      if (!verifier.verify(publicKey.replaceAll("\\n", "\n"), signature, "base64")) {
        throw new Error("Signature mismatch");
      }
    },
    (cause): KickWebhookError => ({
      type: "invalid kick webhook",
      detail: "Kick webhook signature is invalid",
      cause,
    }),
  )();
}

export class KickWebhookHandler {
  constructor(
    private readonly publicKey: string,
    private readonly store: ChatStore,
    private readonly broker: ChatEventBroker,
  ) {}

  async handle(headers: Headers, body: string): Promise<NeverthrowResult<void, KickWebhookError>> {
    const publicKey = this.publicKey;
    const store = this.store;
    const broker = this.broker;
    return await safeTry(async function* () {
      const messageId = yield* requiredHeader(headers, "Kick-Event-Message-Id");
      const timestamp = yield* requiredHeader(headers, "Kick-Event-Message-Timestamp");
      const signature = yield* requiredHeader(headers, "Kick-Event-Signature");
      const eventType = yield* requiredHeader(headers, "Kick-Event-Type");
      yield* verifySignature(publicKey, messageId, timestamp, body, signature);

      if (eventType === "chat.message.sent") {
        const message = yield* parseJson(body)
          .andThen((value) => validate(value, KickMessageSchema))
          .mapErr(
            (cause): KickWebhookError => ({
              type: "invalid kick webhook",
              detail: "Kick message payload is invalid",
              cause,
            }),
          );
        const source = await store.getEnabledSourceByProviderId(
          "kick",
          message.broadcaster.user_id,
        );
        if (!source) {
          return erro({ type: "unknown kick source", detail: "Kick source not found" });
        }
        await broker.publish(
          source.userId,
          {
            type: "message",
            message: {
              id: message.message_id,
              sourceId: source.connectedSource.source.sourceId,
              connectionId: source.connectedSource.source.connectionId,
              provider: "kick",
              author: {
                id: message.sender.user_id,
                displayName: message.sender.username,
              },
              text: message.content,
              occurredAt: new Date(message.created_at),
            },
          },
          `kick:${messageId}`,
        );
      }
      return ok(undefined);
    });
  }
}

File: internal/chat/kick_webhook.go
package chat

import (
	"context"
	"crypto"
	"crypto/rsa"
	"crypto/sha256"
	"crypto/x509"
	"encoding/base64"
	"encoding/json"
	"encoding/pem"
	"errors"
	"fmt"
	"net/http"
	"strings"
	"time"
)

const kickWebhookClockSkew = 5 * time.Minute

type KickWebhookError struct {
	Type   string
	Detail string
	Cause  error
}

func (e *KickWebhookError) Error() string { return e.Detail }
func (e *KickWebhookError) Unwrap() error { return e.Cause }

type KickSourceStore interface {
	GetEnabledSourceByProviderID(context.Context, string, string) (*OwnedConnectedSource, error)
}

type KickWebhookHandler struct {
	publicKey *rsa.PublicKey
	store     KickSourceStore
	broker    EventBroker
	now       func() time.Time
}

func NewKickWebhookHandler(publicKey string, store KickSourceStore, broker EventBroker) (*KickWebhookHandler, error) {
	block, _ := pem.Decode([]byte(strings.ReplaceAll(publicKey, `\n`, "\n")))
	if block == nil {
		return nil, errors.New("invalid Kick webhook public key")
	}
	parsed, err := x509.ParsePKIXPublicKey(block.Bytes)
	if err != nil {
		if pkcs1, pkcs1Err := x509.ParsePKCS1PublicKey(block.Bytes); pkcs1Err == nil {
			return &KickWebhookHandler{publicKey: pkcs1, store: store, broker: broker, now: time.Now}, nil
		}
		return nil, err
	}
	key, ok := parsed.(*rsa.PublicKey)
	if !ok {
		return nil, errors.New("Kick webhook public key is not RSA")
	}
	return &KickWebhookHandler{publicKey: key, store: store, broker: broker, now: time.Now}, nil
}

func (handler *KickWebhookHandler) Handle(ctx context.Context, headers http.Header, body string) error {
	messageID, err := requiredKickHeader(headers, "Kick-Event-Message-Id")
	if err != nil {
		return err
	}
	timestamp, err := requiredKickHeader(headers, "Kick-Event-Message-Timestamp")
	if err != nil {
		return err
	}
	signature, err := requiredKickHeader(headers, "Kick-Event-Signature")
	if err != nil {
		return err
	}
	eventType, err := requiredKickHeader(headers, "Kick-Event-Type")
	if err != nil {
		return err
	}
	if err := handler.verify(messageID, timestamp, body, signature); err != nil {
		return err
	}
	if eventType != "chat.message.sent" {
		return nil
	}
	var message struct {
		MessageID   string `json:"message_id"`
		Broadcaster struct {
			UserID   stringOrNumber `json:"user_id"`
			Username string         `json:"username"`
		} `json:"broadcaster"`
		Sender struct {
			UserID   stringOrNumber `json:"user_id"`
			Username string         `json:"username"`
		} `json:"sender"`
		Content   string `json:"content"`
		CreatedAt string `json:"created_at"`
	}
	if err := json.Unmarshal([]byte(body), &message); err != nil || message.MessageID == "" || message.Broadcaster.UserID == "" || message.Broadcaster.Username == "" || message.Sender.UserID == "" || message.Sender.Username == "" {
		return &KickWebhookError{Type: "invalid kick webhook", Detail: "Kick message payload is invalid", Cause: err}
	}
	occurredAt, err := time.Parse(time.RFC3339Nano, message.CreatedAt)
	if err != nil {
		return &KickWebhookError{Type: "invalid kick webhook", Detail: "Kick message payload is invalid", Cause: err}
	}
	source, err := handler.store.GetEnabledSourceByProviderID(ctx, "kick", string(message.Broadcaster.UserID))
	if err != nil {
		return err
	}
	if source == nil {
		return &KickWebhookError{Type: "unknown kick source", Detail: "Kick source not found"}
	}
	normalized := Message{ID: message.MessageID, SourceID: source.ConnectedSource.Source.SourceID, ConnectionID: source.ConnectedSource.Source.ConnectionID, Provider: "kick", Author: Author{ID: string(message.Sender.UserID), DisplayName: message.Sender.Username}, Text: message.Content, OccurredAt: occurredAt}
	return handler.broker.Publish(ctx, source.UserID, StreamEvent{Type: "message", Message: &normalized}, "kick:"+messageID)
}

func (handler *KickWebhookHandler) verify(messageID, timestamp, body, signature string) error {
	occurredAt, err := time.Parse(time.RFC3339Nano, timestamp)
	if err != nil || handler.now().Sub(occurredAt).Abs() > kickWebhookClockSkew {
		return &KickWebhookError{Type: "invalid kick webhook", Detail: "Kick webhook timestamp is stale", Cause: err}
	}
	decoded, err := base64.StdEncoding.DecodeString(signature)
	if err != nil {
		return &KickWebhookError{Type: "invalid kick webhook", Detail: "Kick webhook signature is invalid", Cause: err}
	}
	digest := sha256.Sum256([]byte(messageID + "." + timestamp + "." + body))
	if err := rsa.VerifyPKCS1v15(handler.publicKey, crypto.SHA256, digest[:], decoded); err != nil {
		return &KickWebhookError{Type: "invalid kick webhook", Detail: "Kick webhook signature is invalid", Cause: err}
	}
	return nil
}

func requiredKickHeader(headers http.Header, name string) (string, error) {
	value := headers.Get(name)
	if value == "" {
		return "", &KickWebhookError{Type: "invalid kick webhook", Detail: fmt.Sprintf("%s is missing", name)}
	}
	return value, nil
}
